File: frontend/src/utils/continuityRiskEngine.ts
import type { ContinuityRiskLevelT } from "@/models/careEpisode";

/**
 * CONTINUITY risk only. "How likely is this patient's care journey to
 * break, and what should be done about it?" This is the engine behind
 * Referral Rescue (canonical context Section 8) — explainable,
 * rule-based, and always paired with a recommended action, never a bare
 * score. A patient can be clinically low-risk and continuity-high-risk
 * at the same time; this file never looks at symptoms/vitals directly,
 * only at journey signals.
 */

interface ContinuityRiskInput {
  referralAcknowledged: boolean;
  referralOverdueSla: boolean;
  priorMissedFollowUp: boolean;
  longTravelDistance: boolean; // e.g. >20km to nearest appropriate facility
  poorConnectivityArea: boolean;
  receivingFacilityOverloaded: boolean;
  chronicCondition: boolean;
}

export interface ContinuityRiskResult {
  score: number;
  level: ContinuityRiskLevelT;
  reasons: string[];
  recommendedAction?: string;
}

export function computeContinuityRisk(input: ContinuityRiskInput): ContinuityRiskResult {
  const reasons: string[] = [];
  let score = 0;

  if (!input.referralAcknowledged) {
    score += 2;
    reasons.push("Referral not yet acknowledged");
  }
  if (input.referralOverdueSla) {
    score += 3;
    reasons.push("Referral not acknowledged within SLA");
  }
  if (input.priorMissedFollowUp) {
    score += 2;
    reasons.push("Previous follow-up was missed");
  }
  if (input.longTravelDistance) {
    score += 1;
    reasons.push("Long travel distance to appropriate facility");
  }
  if (input.poorConnectivityArea) {
    score += 1;
    reasons.push("Patient's area has poor connectivity");
  }
  if (input.receivingFacilityOverloaded) {
    score += 2;
    reasons.push("Receiving facility currently overloaded");
  }
  if (input.chronicCondition) {
    score += 1;
    reasons.push("Chronic condition requiring continued monitoring");
  }

  const level: ContinuityRiskLevelT = score >= 6 ? "high" : score >= 3 ? "medium" : "low";

  let recommendedAction: string | undefined;
  if (level === "high") {
    recommendedAction = "Escalate referral to supervisor";
  } else if (level === "medium") {
    recommendedAction = "Notify referring worker to follow up";
  }

  return { score, level, reasons, recommendedAction };
}

File: backend/app/services/continuity_risk_engine.py
from app.core.config import settings
from app.models.continuity import ContinuityRiskLevel
from app.schemas.continuity import ContinuityRiskScoreRequest, ContinuityRiskScoreOut

"""
CONTINUITY risk only — "how likely is this patient's care journey to
break, and what should be done about it?" Mirrors the frontend's
src/utils/continuityRiskEngine.ts. Explainable and rule-based, always
paired with a recommended action, never a bare score — this is the
engine behind Referral Rescue (canonical context Section 8). Never
looks at symptoms/vitals directly, only journey signals.
"""


def compute_continuity_risk(
    payload: ContinuityRiskScoreRequest,
) -> tuple[int, ContinuityRiskLevel, list[str], str | None]:
    reasons: list[str] = []
    score = 0

    if not payload.referral_acknowledged:
        score += 2
        reasons.append("Referral not yet acknowledged")
    if payload.referral_overdue_sla:
        score += 3
        reasons.append("Referral not acknowledged within SLA")
    if payload.prior_missed_follow_up:
        score += 2
        reasons.append("Previous follow-up was missed")
    if payload.long_travel_distance:
        score += 1
        reasons.append("Long travel distance to appropriate facility")
    if payload.poor_connectivity_area:
        score += 1
        reasons.append("Patient's area has poor connectivity")
    if payload.receiving_facility_overloaded:
        score += 2
        reasons.append("Receiving facility currently overloaded")
    if payload.chronic_condition:
        score += 1
        reasons.append("Chronic condition requiring continued monitoring")

    if score >= settings.continuity_risk_high_threshold:
        level = ContinuityRiskLevel.HIGH
    elif score >= settings.continuity_risk_medium_threshold:
        level = ContinuityRiskLevel.MEDIUM
    else:
        level = ContinuityRiskLevel.LOW

    recommended_action: str | None = None
    if level == ContinuityRiskLevel.HIGH:
        recommended_action = "Escalate referral to supervisor"
    elif level == ContinuityRiskLevel.MEDIUM:
        recommended_action = "Notify referring worker to follow up"

    return score, level, reasons, recommended_action
